app/code/Magento/Sitemap/Model/ProductSitemapItemResolver.php

```diff
@@ -0,0 +1,108 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\Sitemap\Model;
+
+use Magento\Framework\App\Config\ScopeConfigInterface;
+use Magento\Sitemap\Model\ResourceModel\Catalog\ProductFactory;
+use Magento\Store\Model\ScopeInterface;
+
+class ProductSitemapItemResolver implements SitemapItemResolverInterface
+{
+    /**#@+
+     * Xpath config settings
+     */
+    const XML_PATH_PRODUCT_CHANGEFREQ = 'sitemap/product/changefreq';
+    const XML_PATH_PRODUCT_PRIORITY = 'sitemap/product/priority';
+    /**#@-*/
+
+    /**
+     * Product factory
+     *
+     * @var ProductFactory
+     */
+    private $productFactory;
+
+    /**
+     * Sitemap item factory
+     *
+     * @var SitemapItemInterfaceFactory
+     */
+    private $itemFactory;
+
+    /**
+     * Scope config
+     *
+     * @var ScopeConfigInterface
+     */
+    private $scopeConfig;
+
+    /**
+     * ProductSitemapItemResolver constructor.
+     *
+     * @param ScopeConfigInterface $scopeConfig
+     * @param ProductFactory $productFactory
+     * @param SitemapItemInterfaceFactory $itemFactory
+     */
+    public function __construct(
+        ScopeConfigInterface $scopeConfig,
+        ProductFactory $productFactory,
+        SitemapItemInterfaceFactory $itemFactory
+    ) {
+        $this->scopeConfig = $scopeConfig;
+        $this->productFactory = $productFactory;
+        $this->itemFactory = $itemFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getItems($storeId)
+    {
+        $collection = $this->productFactory->create()->getCollection($storeId);
+        $items = array_map(function ($item) use ($storeId) {
```

Comment:
Does we need exclude disabled products also? And this code will not work on large collections